# Comparing Hacker News Posts: Which Receives More Comments

Description: In this project we will explore Hacker News, a popular website started by the startup incubator Y Combinator, where user-submitted stories are voted and commented on, similar to reddit. Specifically, we will compare two types of posts to determine the following:
1. Do Ask HN or Show HN receive more comments on average?
2. Do posts created at a certain time receive more comments on average?

## Analyzing Data Prior to Mining
To begin, we read the data in to get an idea of what our data actually looks like.

In [3]:
#Read in data

import csv
hnfile = open("hacker_news.csv")
hn = list(csv.reader(hnfile))

hn[:5]

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'],
 ['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01']]

## Removing Headers from a List of Lists

In [4]:
#Removing headers
headers = hn[0]
hn = hn[1:]
print(headers)
hn[:5]

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879',
  'Interactive Dynamic Video',
  'http://www.interactivedynamicvideo.com/',
  '386',
  '52',
  'ne0phyte',
  '8/4/2016 11:52'],
 ['10975351',
  'How to Use Open Source and Shut the Fuck Up at the Same Time',
  'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/',
  '39',
  '10',
  'josep2',
  '1/26/2016 19:30'],
 ['11964716',
  "Florida DJs May Face Felony for April Fools' Water Joke",
  'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/',
  '2',
  '1',
  'vezycash',
  '6/23/2016 22:20'],
 ['11919867',
  'Technology ventures: From Idea to Enterprise',
  'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429',
  '3',
  '1',
  'hswarna',
  '6/17/2016 0:01'],
 ['10301696',
  'Note by Note: The Making of Steinway L1037 (2007)',
  'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0',
  '8',
  '2',
  'walterbell',
  '9/30/2015 4:12']]

As shown, the data set above contains the 'id', 'title', 'url', number of upvotes as 'num_points, number of comments on the post as 'num_comments', 'author', and the date/time it was created as 'created_at'.

## Extracting the Ask HN and Show HN Posts

In [5]:
#Identifying the 
ask_posts = []
show_posts = []
other_posts = []

for post in hn:
    title = post[1]
    if title.lower().startswith("ask hn"):
        ask_posts.append(post)
    elif title.lower().startswith("show hn"):
        show_posts.append(post)
    else:
        other_posts.append(post)

print(len(ask_posts))
print(len(show_posts))
print(len(other_posts))
    

1744
1162
17194


## Calulating the Average Number of comments for Ask HN & Show HN posts. 

In [6]:
total_ask_comments = 0
for post in ask_posts:
    num_cmmts = int(post[4])
    total_ask_comments += num_cmmts
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)

total_show_comments = 0
for post in show_posts:
    num_cmts = int(post[4])
    total_show_comments += num_cmts
avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

14.038417431192661
10.31669535283993


On average ASK HN posts receive more comments (14.04) than SHOW HN posts (10.32). 

## Focusing on Ask HN Posts
Since ask hn posts are more likely to receive more comments, we will focus our remaining analysis on these posts.

## Determining if ask posts created at a certain time are more likely to attract comments.

In [7]:
#Calculating the amount of ask posts and comments by hour created.

import datetime as dt
result_list = []

for post in ask_posts:
    created = post[6]
    comments = int(post[4])
    result_list.append([created, comments])

counts_by_hour = {}
comments_by_hour = {}

for result in result_list:
    date = result[0]
    comment = result[1]
    dformat = "%m/%d/%Y %H:%M"
    time_dt = dt.datetime.strptime(date, dformat).strftime("%H")
    
    if time_dt not in counts_by_hour:
        counts_by_hour[time_dt] = 1
        comments_by_hour[time_dt] = comment
    else:
        counts_by_hour[time_dt] += 1
        comments_by_hour[time_dt] += comment
print(counts_by_hour)
print(comments_by_hour)

{'14': 107, '00': 55, '12': 73, '01': 60, '22': 71, '19': 110, '18': 109, '06': 44, '13': 85, '23': 68, '05': 46, '17': 100, '09': 45, '02': 58, '07': 34, '16': 108, '10': 59, '04': 47, '08': 48, '11': 58, '15': 116, '21': 109, '20': 80, '03': 54}
{'14': 1416, '00': 447, '12': 687, '01': 683, '22': 479, '19': 1188, '18': 1439, '06': 397, '13': 1253, '23': 543, '05': 464, '17': 1146, '09': 251, '02': 1381, '07': 267, '16': 1814, '10': 793, '04': 337, '08': 492, '11': 641, '15': 4477, '21': 1745, '20': 1722, '03': 421}


## Calculating the Average Number of Comments per Post for Each Hour

In [8]:
# we'll use these two dictionaries to calculate the average number of comments for posts created during each hour of the day.

avg_by_hour = []
for hour in comments_by_hour:
    avg_by_hour.append([hour, comments_by_hour[hour] / counts_by_hour[hour]])

avg_by_hour    

[['14', 13.233644859813085],
 ['00', 8.127272727272727],
 ['12', 9.41095890410959],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['19', 10.8],
 ['18', 13.20183486238532],
 ['06', 9.022727272727273],
 ['13', 14.741176470588234],
 ['23', 7.985294117647059],
 ['05', 10.08695652173913],
 ['17', 11.46],
 ['09', 5.5777777777777775],
 ['02', 23.810344827586206],
 ['07', 7.852941176470588],
 ['16', 16.796296296296298],
 ['10', 13.440677966101696],
 ['04', 7.170212765957447],
 ['08', 10.25],
 ['11', 11.051724137931034],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['03', 7.796296296296297]]

# Producing Top 5 hours of the day with the highest average number of comments 

In [17]:
# create a list the equals avg_by_hour with swapped columns

swap_avg_by_hour = []
for hour in avg_by_hour:
    swap_avg_by_hour.append([hour[1],hour[0]])

#swapped column list 
swap_avg_by_hour
print('\n')

#sorted swapped by desc
sorted_swap = sorted(swap_avg_by_hour, reverse=True)
sorted_swap

print("Top 5 Hours for Ask Posts Comments")
for avg in sorted_swap[:5]:
    hour = dt.datetime.strptime(avg[1],"%H").strftime("%H:%M")
    
    print("{}: {:.2f} average comments per post".format(hour,avg[0]))
    



Top 5 Hours for Ask Posts Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post


# Conclusion

Based on the data above, the top five hours to create post to have a higher chance of receiving comments are:
    1. 3:00PM
    2. 2:00AM
    3. 8:00PM
    4. 4:00PM
    5. 9:00PM
